test/contracts/HCElection.aes

```diff
@@ -9,10 +9,13 @@ contract interface MainStaking =
 
 main contract HCElection =
   record epoch_info =
-    { start                : int,
-      length               : int,
-      seed                 : option(bytes()),
-      staking_distribution : option(list(address * int))
+    { start                 : int,
+      length                : int,
+      seed                  : option(bytes()),
+      staking_distribution  : option(list(address * int)),
+      base_pin_reward       : int,
+      cur_pin_reward        : int,
+      carry_over_pin_reward : int
```

Comment:
Do we really need to keep this in epoch info? We don't expect it to change often, right? Won't it be enough to keep a `pin_reward_per_epoch` and a `current_pin_reward` and at the end of epoch we either reset or accumulate 🤔 This looks a little bit like overkill...

Reply:
`carry_over` seems to be a given for the epoch, but I can definitely see the case for the other two being in the state. What this way _does_ do is that it keeps the number of calls needed to the contract down: everything is fetched once per validation (or even epoch, although I don't think `epoch_info` is cached). I'm open for suggestions. There is also future voting procedures to take into account. And I guess just because it's called init_epochs doesn't mean we can init other things in the state as well?

Reply:
As I said in the comment, my preferred way of doing it would be for it to be dynamically included in the epoch_info call (as it is relevant for the epoch _and_ to keep contract calls down), so pointers to solve that will be muchos appreciar.

Reply:
My point was that you don't seem to have different values in different epochs!? `carry_over` isn't known for next-next epoch, and surely not for next-next-next epoch... At the end of each epoch these numbers needs to be adjusted, but you don't know them several epochs in advance 🤔

Reply:
Correct. Again, the main argument for keeping them with the epoch is to simplify getting to epoch related state (which they truly are, it’s just that they can’t be calculated in advance like some of the other ones as you point out). This is useful for both the end-of-epoch validation and also for MDW, who don’t need to assemble this from multiple end-points. 

I don’t have particular love for either, I just want something that is efficient, usable and reasonably elegant. Perhaps @ThomasArts or @happi  can chime in?

Reply:
Is _resource_ efficiency (i.e. cost of calling the contract) actually an issue. My main pro about current solution is developer efficiency?

Reply:
My suggestion is we leave this current (working, but slightly contentious) implementation in, and if we either see issues with it or just can't stand it given a couple of days I will happily rewrite it?

Reply:
Consensus on how to do this:
1. Move pin fields into it's own (record?) fields on the state
2. Create accessors for the pin state (`pin_info`) on the contract and in `aec_chain_ch.erl`
3. change `aec_consensus_ch.erl`code to use the new `pin_info`.